homeassistant/components/minecraft_server/__init__.py

```diff
@@ -0,0 +1,342 @@
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+from homeassistant.util.dt import utcnow
+
+from .const import DOMAIN, MANUFACTURER, SIGNAL_NAME_PREFIX
+
+PLATFORMS = ["binary_sensor", "sensor"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up the Minecraft Server component."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, config_entry: ConfigEntry) -> bool:
+    """Set up Minecraft Server from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # Create and store server instance.
+    unique_id = config_entry.unique_id
+    server = MinecraftServer(hass, unique_id, config_entry.data)
+    hass.data[DOMAIN][unique_id] = server
+    await server.async_update()
+
+    # Set up platform(s).
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, platform)
+        )
+
+    return True
+
+
+async def async_unload_entry(
+    hass: HomeAssistantType, config_entry: ConfigEntry
+) -> bool:
+    """Unload Minecraft Server config entry."""
+    unique_id = config_entry.unique_id
+    server = hass.data[DOMAIN][unique_id]
+
+    # Unload platforms.
+    for platform in PLATFORMS:
+        await hass.config_entries.async_forward_entry_unload(config_entry, platform)
+
+    # Clean up.
+    server.stop_periodic_update()
+    hass.data[DOMAIN].pop(unique_id)
+
+    return True
+
+
+class MinecraftServer:
+    """Representation of a Minecraft server."""
+
+    # Private constants
+    _RETRIES_PING = 3
+    _RETRIES_STATUS = 3
+
+    def __init__(
+        self, hass: HomeAssistantType, unique_id: str, config_data: ConfigType
+    ) -> None:
+        """Initialize server instance."""
+        self._hass = hass
+        self._unique_id = unique_id
+
+        # Server data
+        self._name = config_data[CONF_NAME]
+        self._host = config_data[CONF_HOST]
+        self._port = config_data[CONF_PORT]
+        self._scan_interval = config_data[CONF_SCAN_INTERVAL]
+        self._server_online = False
+        self._stop_periodic_update = None
+
+        _LOGGER.debug(
+            "Initializing '%s' (host='%s', port=%s, scan_interval=%s)...",
+            self._name,
+            self._host,
+            self._port,
+            self._scan_interval,
+        )
+
+        # 3rd party library instance
+        self._mcstatus = MCStatus(self._host, self._port)
+
+        # Data provided by 3rd party library
+        self._description = STATE_UNKNOWN
+        self._version = STATE_UNKNOWN
+        self._protocol_version = STATE_UNKNOWN
+        self._latency = STATE_UNKNOWN
+        self._players_online = STATE_UNKNOWN
+        self._players_max = STATE_UNKNOWN
```

Comment:
Seen in an other PR yesterday : do not set states to `STATE_UNKNOWN`, but `None` instead, HA will take care of the rest.